### PR TITLE
[ESIMD] Introduce Masked Compress load/store API

### PR DESCRIPTION
**Pull request by fineg74**

### CODE REVIEW

File: sycl/include/sycl/ext/intel/esimd/memory.hpp [hunk 1]
```diff
@@ -14302,6 +14302,254 @@ named_barrier_signal(uint8_t barrier_id, uint8_t producer_consumer_mode,
...
+/// properties are used.
+///
+template <typename Tx, int N, typename AccessorTy,
+          typename PropertyListT = oneapi::experimental::empty_properties_t>
+__ESIMD_API std::enable_if_t<
+    ext::oneapi::experimental::is_property_list_v<PropertyListT> &&
+        detail::is_accessor_with_v<AccessorTy,
+                                   detail::accessor_mode_cap::can_read>,
+    simd<Tx, N>>
+mask_compress_load(AccessorTy acc, uint32_t global_offset, simd_mask<N> mask,
```
Comment (vmustya) [inline sycl/include/sycl/ext/intel/esimd/memory.hpp:14427]: AFAIR, both LLVM and CPU hardware support [expand load](https://llvm.org/docs/LangRef.html#llvm-masked-expandload-intrinsics) operations, that should be emulated by conditional gather. What is the idea of "compress load"?

File: sycl/include/sycl/ext/intel/esimd/memory.hpp [hunk 2]
```diff
@@ -14302,6 +14302,211 @@ named_barrier_signal(uint8_t barrier_id, uint8_t producer_consumer_mode,
...
+/// properties are used.
+///
+template <typename Tx, int N,
+          typename PropertyListT = oneapi::experimental::empty_properties_t>
+__ESIMD_API std::enable_if_t<
+    ext::oneapi::experimental::is_property_list_v<PropertyListT>, simd<Tx, N>>
+mask_expand_load(Tx *p, simd_mask<N> mask, PropertyListT props = {}) {
+  typedef typename detail::GenerateCompressedBitmask<N>::value offsets;
+  simd<uint32_t, N> offset =
+      cbit(simd<uint32_t, N>(offsets::value) & pack_mask(mask));
```
Comment (sarnex) [inline sycl/include/sycl/ext/intel/esimd/memory.hpp:14344]: do you mind adding comments that explain this logic, i don't understand it completely just from reading, thanks

File: sycl/include/sycl/ext/intel/esimd/memory.hpp [hunk 3]
```diff
@@ -14302,6 +14302,211 @@ named_barrier_signal(uint8_t barrier_id, uint8_t producer_consumer_mode,
...
+/// for (int i = 0; i < N; ++i) {
+///       if (Mask[i]) {
+///           Result[i] = *(p + Index);
+///           ++Index;
+///       }
+/// }
+/// \endcode
+///
+///
+/// @tparam Tx is element type.
```
Comment (sarnex) [inline sycl/include/sycl/ext/intel/esimd/memory.hpp:14330]:
```suggestion
/// @tparam Tx is the element type.
```

File: sycl/include/sycl/ext/intel/esimd/memory.hpp [hunk 4]
```diff
@@ -14302,6 +14302,211 @@ named_barrier_signal(uint8_t barrier_id, uint8_t producer_consumer_mode,
...
+/// @tparam N is the data size
+/// @param p is the base address for this operation.
+/// @param mask is the mask determining which elements will be read
+/// @param props The compile-time properties. Only cache hint
+/// properties are used.
```
Comment (sarnex) [inline sycl/include/sycl/ext/intel/esimd/memory.hpp:14335]: nit: mix between having a `.` at the end and not, i think we should have `.` to match other APIs

File: sycl/include/sycl/ext/intel/esimd/memory.hpp [hunk 5]
```diff
@@ -14302,6 +14302,211 @@ named_barrier_signal(uint8_t barrier_id, uint8_t producer_consumer_mode,
...
+///
+template <typename Tx, int N,
+          typename PropertyListT = oneapi::experimental::empty_properties_t>
+__ESIMD_API std::enable_if_t<
+    ext::oneapi::experimental::is_property_list_v<PropertyListT>, simd<Tx, N>>
+mask_expand_load(Tx *p, simd_mask<N> mask, PropertyListT props = {}) {
```
Comment (sarnex) [inline sycl/include/sycl/ext/intel/esimd/memory.hpp:14341]: can `T*` be `const T*` for the load APIs?

File: sycl/test-e2e/ESIMD/mask_compress_store.cpp [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+      mask_compress_store(Result, 0, Input, Mask);
+      mask_compress_store(OutputAcc_out, 0, Input, Mask);
+      Result.copy_to(OutputVectorPtr);
+    });
+    cgh.single_task(Kernel);
+  });
+  Queue.wait();
+
+  for (int I = 0; I < N; I++) {
+    if (Output[I] != OutputVector[I] && OutputAcc[I] != Output[I] &&
```
Comment (sarnex) [inline sycl/test-e2e/ESIMD/mask_compress_store.cpp:59]: Do we need to change the logic from `&&` to `||` to handle the case where the acc api works but the usm one fails?

File: sycl/test-e2e/ESIMD/mask_compress_store.cpp [hunk 7]
```diff
@@ -0,0 +1,86 @@
...
+int main() {
+
+  bool Pass = true;
+  auto Q = queue{gpu_selector_v};
+  esimd_test::printTestLabel(Q);
+  Pass &= test<8>(Q);
+  Pass &= test<16>(Q);
+  Pass &= test<32>(Q);
+
+  if (Pass) {
```
Comment (sarnex) [inline sycl/test-e2e/ESIMD/mask_compress_store.cpp:82]: nit: remove `{}`

**Review by sarnex — COMMENTED** [review]
inital comments, thanks!

File: sycl/include/sycl/ext/intel/esimd/memory.hpp [hunk 8]
```diff
@@ -14302,6 +14302,211 @@ named_barrier_signal(uint8_t barrier_id, uint8_t producer_consumer_mode,
...
+///
+/// @tparam Tx is element type.
+/// @tparam N is the data size
+/// @tparam M is the source data size
+/// @param dst is the vector to write to.
+/// @param global_offset is an offset to use for all writes.
+/// @param vals is the data to store
+/// @param mask is the mask determining which elements will be stored
+///
+template <typename Tx, int N, int M>
```
[the remark this answers is not in the document]
Comment (sarnex) [reply sycl/include/sycl/ext/intel/esimd/memory.hpp]: got it, thanks

**Review by sarnex — COMMENTED** [review]
second pass below, only minor comments, thanks!

File: sycl/include/sycl/ext/intel/esimd/memory.hpp [hunk 9]
```diff
@@ -14302,6 +14302,241 @@ named_barrier_signal(uint8_t barrier_id, uint8_t producer_consumer_mode,
...
+  // contains i 1's i.e. 0,1,3,7,...
+  using offsets = typename detail::GenerateCompressedBitmask<N>::value;
+  // Performing '&' operation with packed mask will leave at index i a bitmask
+  // with number of 1's corresponding to a number of elements to be loaded so
+  // far (number of 1's in the mask preceding the index i). Number of 1's
+  // becomes an index for compressed store/expanded load operation.
+  simd<uint32_t, N> offset =
+      cbit(simd<uint32_t, N>(offsets::value) & pack_mask(mask));
+
+  simd<uint16_t, N> Indices = global_offset + offset;
```
[the remark this answers is not in the document]
Comment (sarnex) [reply sycl/include/sycl/ext/intel/esimd/memory.hpp:14535]: ah sorry, i thought they were memory offsets

**Review by sarnex — APPROVED** [review]
thanks!

**Pull request merged**